### DISCUSSION TITLE
ImapClient should preserve user name, password and host in self when successfully logged in

### DISCUSSION THREAD

**Discussion by linuxmagic-mp** [mjs/imapclient#506] [Ideas]
Think something is generally missing, when login is called, whether manually, or using an INI, there should be a self.authenticated, self.username, self.host etc.. would make it easier for other functions to simply access self to get these values.. Does this make sense? Maybe even set it when not authenticated..

**Comment by mjs** [comment]
I'm open to storing the username and hostname on the IMAPClient instance but am a little reluctant to store the password. Keeping secrets around in memory is generally a bad idea from a security standpoint.

What's the use case for this? It's not like your application couldn't track these details itself alongside with the IMAPClient instance.

**Reply by linuxmagic-mp** [reply]
While in 'general' I agree with you, however that in general is the behavior of most email clients, and some clients will make multiple connections for servicing the same email account for instance.  The username and hostname though can be elements of other actions, that are specific to an account/hostname, eg if we want a setting specific to that.  Example use case would be when a function using the library needs to look up configuration items that are specific to a username/hostname, it can be a simple follow up call to check/use $self.username for locating locally stored account specific settings.

**Reply by linuxmagic-mp** [reply]
oops.. <ctrl><enter>.. continuing.. my use case is needing to set the username and hostname PRIOR to the login call, to be able to add a function prior to login, that will be able to retrieve the 2FA token needed for CLIENTID capabilitiies if advertised, based on the username/hostname that is set, if the capabilities are advertised.

**Reply by NicolasLM** [reply]
Hey, while I am not exactly sure I understand what you are trying to achieve, I wanted to remind you that you can extend IMAPClient very easily to implement your custom behavior. Unless the change you are proposing clearly benefits others, I would go this route:


```python
from imapclient import IMAPClient

class CustomIMAPClient(IMAPClient):

    def __init__(self, custom_config, *args, **kwargs):
        self.custom_config = custom_config
        super().__init__(*args, **kwargs)
        
with CustomIMAPClient("my-custom-config", host="imap.host.org") as client:
    client.login('someone', 'secret')
```
